rest-api/offline/table_exporter.py

```diff
@@ -1,33 +1,114 @@
+import hashlib
+import random
 import re
+import struct
 
+from dao.database_factory import get_database
 from google.appengine.api import app_identity
 from google.appengine.ext import deferred
 from offline.sql_exporter import SqlExporter
 from werkzeug.exceptions import BadRequest
 
 _TABLE_PATTERN = re.compile("^[A-Za-z0-9_]+$")
 
+# TODO(calbach): Factor this out into the datastore config.
+_DEIDENTIFY_DB_TABLE_WHITELIST = {
+  'rdr': set([
+      'participant_view',
+      'physical_measurements_view',
+      'questionnaire_response_answer_view'
+  ])
+}
+
+
 class TableExporter(object):
   """API that exports data from our database to UTF-8 CSV files in GCS.
 
   Used instead of Cloud SQL export because it handles newlines and null characters in a way that
   other CSV clients (e.g. BigQuery, Google Sheets) can actually understand.
   """
 
+  @staticmethod
+  def _obfuscate_id(pmi_id, salt):
+    """
+    One-way ID obfuscation, intended for PMI participant IDs.
+
+    This method should aim to (1) avoid hash collisions across PMI ID inputs and (2) not be
+    reversible or rerunnable by anyone without RDR production data access.
+
+    Params:
+      - pmi_id (number) must be an integer participant ID
+      - salt (number) must be a long
+
+    Returns: a positive integer
+    """
+    h = hashlib.sha1()
+    h.update(struct.pack('>l', pmi_id))
+    h.update(salt)
+    # Just take the first 4 bytes so that the output ID is an integer, roughly in the same domain as
+    # the input PMI participant ID.
+    b = h.digest()[0:4]
+    return abs(struct.unpack('>l', b)[0])
+
   @classmethod
-  def _export_csv(cls, bucket_name, database, directory, table_name):
+  def _export_csv(cls, bucket_name, database, directory, deidentify_salt, table_name):
     assert _TABLE_PATTERN.match(table_name)
     assert _TABLE_PATTERN.match(database)
-    SqlExporter(bucket_name, use_unicode=True).run_export('%s/%s.csv' % (directory, table_name),
-                                                          'SELECT * FROM %s.%s' %
-                                                          (database, table_name))
+
+    transformf = None
+    if deidentify_salt:
+      # Deidentification requested: hash outgoing participant IDs with a consistent salt across this
+      # export. Cache obfuscated participant IDs across row callbacks to avoid recomputation and to
+      # detect collisions.
+      pmi_to_obfuscated = {}
+      obfuscated_to_pmi = {}
+      def f(row_proxy):
+        out = [v for v in row_proxy]
+        for i, key in enumerate(row_proxy.keys()):
+          if key != 'participant_id':
+            continue
+
+          pmi_id = out[i]
+          if pmi_id not in pmi_to_obfuscated:
+            obf_id = TableExporter._obfuscate_id(pmi_id, deidentify_salt)
+            pmi_to_obfuscated[pmi_id] = obf_id
```

Comment:
One cause for concern: how much memory this map takes up when it has 1M entries in it. Assuming you switch the values to longs, the keys + values will be 16M bytes, but the overhead of map entries could be significant.

Reply:
I'm not a python performance expert, but my guess is that I'd rather take the memory hit than the processing time.

This cache is not critical though, so I'm fine to remove it (though it does have a nice benefit that we can detect collisions, in the event of a programming error here). I don't necessarily trust this scheme if we scale it up to 1m participants (e.g. the issues you found in your comments).

Perhaps leave for now, then see what the performance looks like in sandbox (if that's in a good state?)